### PR TITLE
[1.12] added block whitelist for covers

### PR DESCRIPTION
**Pull request by SanAndreaP**
enables adding blocks that are not considered full cubes (like glass)
defaults to add glass (regular, stained and TF hardened)

### CODE REVIEW

**Comment by KingLemming** [comment]
@covers1624 - you wanted to handle this differently. Please elaborate.

**Comment by covers1624** [comment]
From what i remember, specific transparent models had issues rendering, notably leaves (probably a few others), ideally a blacklist should be implemented and the issues solved, i'ma fiddle with the rendering a little and see if i can resolve it then ill get back to this.

File: src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java [hunk 1]
```diff
@@ -1,16 +1,34 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 public class CoverHelper {
+	public static final List<String> BLOCK_WHITELIST = new ArrayList<>();
+
+	public static boolean preInit() {
+		String[] whitelist = {"minecraft:glass", "thermalfoundation:glass:32767", "minecraft:stained_glass:32767"};
```
Comment (covers1624) [inline src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java:24]: I honestly don't like this format for some reason, there has to be a standard. Maybe a separate json file should be used.  
Example:  
```json
[
	{
		//Meta optional, specifies only meta 0.
		"block" : "minecraft:glass"
	},
	{
		"block": "thermalfoundation:glass",
		"meta": false // False specifies ore dict, all meta's ever. (can be int to specify specific meta)
	},
	{
		"block": "minecraft:stained_glass",
		"states": {
			"color": {
				"whitelist": [ // Whitelist specifies these /can/ exist but must not collide with the blacklist.
					"white",
					"red",
					"light_blue"
				],
				"blacklist": [ //These values can't exist ever.
					"lime",
					"gray",
					"green"
				]
			}
		}
	},
	{
		"block": "minecraft:some_other_block",
		"states": {
			"color": [ // Can also be an array of pure whitelist.
				"white",
				"red",
				"light_blue"
			]
		}
	}
]
```

Honestly probably a little overkill, even without the state checking..

File: src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java [hunk 2]
```diff
@@ -34,7 +52,17 @@ public static boolean isValid(Block block, int meta) {
...
+			if( !(block.hasTileEntity(state) || block.hasTileEntity()) ) {
+				if( state.isFullCube() ) {
+					return true;
+				} else {
+					ResourceLocation wlKey = Block.REGISTRY.getNameForObject(block);
+					return BLOCK_WHITELIST.contains(wlKey.toString())
+								   || BLOCK_WHITELIST.contains(String.format("%s:%d", wlKey, meta))
+								   || BLOCK_WHITELIST.contains(String.format("%s:%d", wlKey, OreDictionary.WILDCARD_VALUE));
+				}
+			}
```
Comment (covers1624) [inline src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java:64]: Formatting is a little off from everything else here..

**Review by covers1624 — COMMENTED** [review]
Looks good for a whitelist, don't really like the format for the config (suggested alt's). Minor formatting.

File: src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java (hunk 1 above)
Comment (SanAndreaP) [reply src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java:24]: I was going for a CraftTweaker-like format (they have ```<domain:item:meta>``` there). I could make it JSON if desired.
The blacklist woul only be useful for blocks that are full cubes, no?
I'd like to do it rather like this, if we decide on JSON:
```
{
    "whitelist": [
        {
            "block": "minecraft:stained_glass",
            "states": {
                "color": [
                    "orange",
                    "purple",
                    "blue"
                ]
            }
        }
    ],
    "blacklist": [
        {
            "block": "minecraft:wool",
            "meta": true
        }
    ]
}
```

A User would see immediately what is in the whitelist and what in the blacklist.

File: src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java (hunk 2 above)
Comment (SanAndreaP) [reply src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java:64]: The formatting looks fine on my IDE, it could be because of github.

Comment (covers1624) [reply src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java:64]: No I mean it's off from our formatting.

File: src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java (hunk 1 above)
Comment (covers1624) [reply src/main/java/cofh/thermaldynamics/duct/attachments/cover/CoverHelper.java:24]: Today im looking at the rendering, the idea is to turn it in to a pure blacklist and have all non tiles included by default, no matter how silly it looks.

**Comment by covers1624** [comment]
Ok, managed to get cover item rendering fixed for some blocks. Also implemented a blacklist in json format. Thanks for the effort creating this pr!

**Pull request closed without merging**